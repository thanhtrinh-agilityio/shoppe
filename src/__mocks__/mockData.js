export const PRODUCT_ITEM = {
  id: 1,
  imageUrl:
    'https://firebasestorage.googleapis.com/v0/b/test-1-50371.appspot.com/o/shoppe-products%2FImg-04.png?alt=media&token=641c3cfe-cab6-4498-8632-eb511434e042',
  name: 'Lira Earrings',
  price: 20,
  status: 'Sold out',
}

export const PRODUCT_DETAIL = {
  id: 1,
  imageUrl:
    'https://firebasestorage.googleapis.com/v0/b/test-1-50371.appspot.com/o/shoppe-products%2FImg-04.png?alt=media&token=641c3cfe-cab6-4498-8632-eb511434e042',
  name: 'Lira Earrings',
  price: 20,
  status: 'Sold out',
  quantity: 1,
  introduction:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam placerat, augue a volutpat hendrerit, sapien tortor faucibus augue, a maximus elit ex vitae libero. Sed quis mauris eget arcu facilisis consequat sed eu felis. ',
}

export const LIST_PRODUCT = [
  {
    id: 1,
    imageUrl: '/images/liraEarring.svg',
    name: 'Lira Earrings',
    price: 20,
    status: '-25%',
  },
  {
    id: 2,
    imageUrl: '/images/liraEarring.svg',
    name: 'Lira Earrings',
    price: 20,
    status: '',
  },
  {
    id: 3,
    imageUrl: '/images/liraEarring.svg',
    name: 'Kaede Hair Pin Set Of 3',
    price: 20,
    status: '',
  },
  {
    id: 4,
    imageUrl: '/images/liraEarring.svg',
    name: 'Hair Pin Set of 3',
    price: 20,
  },
  {
    id: 5,
    imageUrl: '/images/liraEarring.svg',
    name: 'Plaine Necklace',
    price: 20,
    status: 'Sold out',
  },
  {
    id: 6,
    imageUrl: '/images/liraEarring.svg',
    name: 'Yuki Hair Pin Set of 3',
    price: 20,
    status: '',
  },
]
