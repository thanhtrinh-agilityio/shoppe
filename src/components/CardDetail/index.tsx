import { memo, useState } from 'react'
import Image, { ImageLoader } from 'next/image'
import Link from 'next/link'
import { Box, Flex, Text } from '@chakra-ui/react'

// Utils
import { currencyFormat } from '@utils/index'

// Types
import { IProductDetail } from '@self-types/index'

// Components
import Button from '@components/Button'
import Quantity from '@components/Quantity'

// Services
import myImageLoader from '@services/imageLoader'

interface CardDetailProps {
  productDetail: IProductDetail
  handleAddCart: (data: IProductDetail) => void
}

const CardDetail = ({ productDetail, handleAddCart }: CardDetailProps) => {
  const { name, price, imageUrl, id, introduction } = productDetail

  const [quantityProduct, setQuantityProduct] = useState<number>(1)

  // Handle increment quantity
  const handleIncrementQuantity = () => setQuantityProduct(quantityProduct + 1)

  // Handle decrement quantity
  const handleDecrementQuantity = () => {
    const value = quantityProduct > 1 ? quantityProduct - 1 : 1
    setQuantityProduct(value)
  }

  // Handle Click button Add to cart
  const handleClickAddCart = () => {
    const newCart: IProductDetail = {
      ...productDetail,
      quantity: quantityProduct,
    }

    handleAddCart(newCart)
  }

  return (
    <Box
      display="flex"
      justifyContent="space-between"
      height="600px"
      maxW="1248px"
    >
      <Box display="flex" gap="78px">
        <Image
          width={560}
          height={600}
          src={imageUrl}
          alt={name}
          loader={myImageLoader as ImageLoader}
        />
        <Flex h="full" direction="column">
          <Link href={`/products/${id}`} passHref>
            <Text
              as="h2"
              fontSize="26px"
              textColor="dark"
              data-testid={`link-to-detail-page/${id}`}
            >
              {name}
            </Text>
          </Link>
          <Text fontSize="20px" mt="24px" textColor="beaver">
            {currencyFormat(price)}
          </Text>
          <Text
            fontSize="16px"
            mt="46px"
            textColor="secondary"
            textAlign="justify"
          >
            {introduction}
          </Text>
          <Flex mt="75px" gap="58px">
            <Quantity
              onIncreaseCartQuantity={handleIncrementQuantity}
              onDecrementCartQuantity={handleDecrementQuantity}
              quantity={quantityProduct}
            />
            <Button
              onClick={handleClickAddCart}
              variant="primary"
              size="default"
            >
              ADD TO CART
            </Button>
          </Flex>
        </Flex>
      </Box>
    </Box>
  )
}

export default memo(CardDetail)
