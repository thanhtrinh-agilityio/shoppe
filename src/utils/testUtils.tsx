import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'

// Themes
import { customTheme } from '@themes/index'

// Contexts
import { CartProvider } from '@contexts/CartProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ChakraProvider theme={customTheme}>
      <CartProvider>{component}</CartProvider>
    </ChakraProvider>,
  )
}

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { customRender as render }
