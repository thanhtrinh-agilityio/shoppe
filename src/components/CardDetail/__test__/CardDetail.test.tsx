// Mocks
import { PRODUCT_DETAIL } from '@mocks/mockData'

// Utils
import { fireEvent, render, screen } from '@utils/testUtils'

// Components
import CardDetail from '@components/CardDetail'

describe('CardDetail render', () => {
  const mockProps = {
    productDetail: PRODUCT_DETAIL,
    handleAddCart: jest.fn(),
  }
  it('Should match data for card component', () => {
    const { container } = render(<CardDetail {...mockProps} />)

    expect(container).toMatchSnapshot()
  })

  it('Should click button add', () => {
    render(<CardDetail {...mockProps} />)

    const buttonAddCart = screen.getByText('ADD TO CART')

    fireEvent.click(buttonAddCart)
    expect(mockProps.handleAddCart).toBeCalled()
  })

  it('Should click increase and decrease the quantity', () => {
    render(<CardDetail {...mockProps} />)

    const increaseButton = screen.getByTestId('increase-button')
    const decreaseButton = screen.getByTestId('decrement-button')

    // Test increase quantity
    fireEvent.click(increaseButton)
    expect(screen.getByTestId('quantity').textContent).toBe('2')

    // Test decrease quantity
    fireEvent.click(decreaseButton)
    expect(screen.getByTestId('quantity').textContent).toBe('1')
  })

  it('Should click reduce the number should not be less than 1', () => {
    render(<CardDetail {...mockProps} />)

    const decreaseButton = screen.getByTestId('decrement-button')

    fireEvent.click(decreaseButton)
    expect(screen.getByTestId('quantity').textContent).toBe('1')
  })
})
