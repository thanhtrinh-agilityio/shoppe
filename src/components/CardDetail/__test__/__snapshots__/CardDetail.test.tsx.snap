// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDetail render Should match data for card component 1`] = `
<div>
  <div
    class="css-120yquv"
  >
    <div
      class="css-1t8dwq6"
    >
      <img
        alt="Lira Earrings"
        data-nimg="1"
        decoding="async"
        height="600"
        loading="lazy"
        src="https://firebasestorage.googleapis.com/v0/b/test-1-50371.appspot.com/o/shoppe-products%2FImg-04.png?alt=media&token=641c3cfe-cab6-4498-8632-eb511434e042?w=1200&q=75"
        srcset="https://firebasestorage.googleapis.com/v0/b/test-1-50371.appspot.com/o/shoppe-products%2FImg-04.png?alt=media&token=641c3cfe-cab6-4498-8632-eb511434e042?w=640&q=75 1x, https://firebasestorage.googleapis.com/v0/b/test-1-50371.appspot.com/o/shoppe-products%2FImg-04.png?alt=media&token=641c3cfe-cab6-4498-8632-eb511434e042?w=1200&q=75 2x"
        style="color: transparent;"
        width="560"
      />
      <div
        class="css-14sa009"
      >
        <h2
          class="chakra-text css-tq06qo"
          data-testid="link-to-detail-page/1"
          href="/products/1"
        >
          Lira Earrings
        </h2>
        <p
          class="chakra-text css-1plr3rm"
        >
          $ 20.00
        </p>
        <p
          class="chakra-text css-ubux72"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aliquam placerat, augue a volutpat hendrerit, sapien tortor faucibus augue, a maximus elit ex vitae libero. Sed quis mauris eget arcu facilisis consequat sed eu felis. 
        </p>
        <div
          class="css-17symod"
        >
          <div
            class="css-drsjqf"
          >
            <button
              aria-label="Search database"
              class="chakra-button css-rzypxw"
              data-testid="increase-button"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <p
              class="chakra-text css-17i4smz"
              data-testid="quantity"
            >
              1
            </p>
            <button
              aria-label="Search database"
              class="chakra-button css-rzypxw"
              data-testid="decrement-button"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  fill="currentColor"
                >
                  <rect
                    height="4"
                    width="20"
                    x="2"
                    y="10"
                  />
                </g>
              </svg>
            </button>
          </div>
          <button
            class="chakra-button css-1a6r7m5"
            type="button"
          >
            ADD TO CART
          </button>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
